src/server/game/Entities/Player/Player.cpp

```diff
@@ -26712,3 +26768,237 @@ std::string Player::GetDebugInfo() const
     sstr << Unit::GetDebugInfo();
     return sstr.str();
 }
+
+// Anticheat System
+void Player::SetUnderACKmount()
+{
+    m_mountTimer = 3000;
+    m_ACKmounted = true;
+}
+
+void Player::SetRootACKUpd(uint32 delay)
+{
+    m_rootUpdTimer = 1500 + delay;
+    m_rootUpd = true;
+}
+
+bool Player::CheckOnFlyHack()
+{
+    if (IsCanFlybyServer())
+        return true;
+
+    if (ToUnit()->IsFalling() || IsFalling())
+        return true;
+
+    if (IsFlying() && !CanFly()) // kick flyhacks
+    {
+        TC_LOG_INFO("anticheat", "Player::CheckMovementInfo :  FlyHack Detected for Account id : %u, Player %s", GetPlayerMovingMe()->GetSession()->GetAccountId(), GetPlayerMovingMe()->GetName().c_str());
+        TC_LOG_INFO("anticheat", "Player::========================================================");
+        TC_LOG_INFO("anticheat", "Player IsFlying but CanFly is false");
+
+        sWorld->SendGMText(LANG_GM_ANNOUNCE_AFH_CANFLYWRONG, GetPlayerMovingMe()->GetName().c_str());
+        return false;
+    }
+
+    if (IsFlying() || IsLevitating() || IsInFlight())
+        return true;
+
+    if (GetTransport() || GetVehicle() || GetVehicleKit())
```

Comment:
If skip check when have vehicle, can we use flyhack without detect it?
Maybe can we get vehicle before and do some check with vehicles?